### PR TITLE
AFE Integration: Submit CSTA form from our server

### PR DESCRIPTION
**Pull request by islemaster**
Adds some code that handles submitting from our server to the CSTA+ form.  Isolating this into a Service because there's some semi-complex translation of the parameters into the appropriate JotForm-friendly format.

I've already added the appropriate configuration values as secrets to AWS secrets manager.

## Links

[AFE Integration TODO list](https://docs.google.com/document/d/1Ny0pTrvkBMnOKx0CqoNnWTWhwq-oCn9IBYekSjiJMBc/edit)

## Testing story

I've added unit tests that check the parameters are passed through into a well-formed format.  I've manually verified that this format is well-formed by submitting matching requests to the CSTA test JotForm and retrieving the generated submissions, confirming that all the data we sent is intact.

I've also verified that the production CSTA form has exactly the same question format as the test CSTA form by asking their REST API for the question metadata on each form and diffing the responses - so it's reasonable to assume that if this feature is working against the test form, it will work against the production form.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/api/v1/amazon_future_engineer_controller.rb [hunk 1]
```diff
@@ -29,7 +29,24 @@ def submit
...
-  rescue Services::AFEEnrollment::Error => e
+
+    # If the teacher requested it, submit to CSTA as well
+    if to_bool(afe_params['csta'])
+      Services::CSTAEnrollment.new.submit(
+        first_name: afe_params['firstName'],
+        last_name: afe_params['lastName'],
+        email: afe_params['email'],
+        school_district_name: '',
+        school_name: '',
```
Comment (islemaster) [inline dashboard/app/controllers/api/v1/amazon_future_engineer_controller.rb]: We're not passing these from the client.  I'm intentionally leaving them unimplemented for now.  In the near future, I think we'll have to add a lookup using NCES id to get this information.

File: config.yml.erb [hunk 2]
```diff
@@ -249,6 +249,8 @@ aws_secret_key:
 
 # Amazon Future Engineer integration
 afe_pardot_form_handler_url:
+csta_jotform_api_key: !Secret
+csta_jotform_form_id: !Secret
```
Comment (islemaster) [inline config.yml.erb:253]: Note: CSTA has provided us with a test form and a live form.  I have all environments pointing at the test form for now.

**Review by hacodeorg — APPROVED** [review]
Looks good based on how much I still remember about JotForm.

**Comment by islemaster** [comment]
Putting this on hold!  I did some more manual testing and found that the submission was not actually working, even though I thought I'd tested all the constituent parts here.  Backing up and finding a different solution...

**Comment by islemaster** [comment]
Okay, this is ready for another look! I've updated my approach, and I think the code is somewhat cleaner now.  We're also getting school info from our records if possible.

File: dashboard/test/lib/services/csta_enrollment_test.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  def failure_response
+    mock.tap do |response|
+      # This is the only real failure I was able to generate when
+      # testing this API - it's a response to sending no fields
+      # at all.
+      response.stubs(:code).returns('400')
+      response.stubs(:body).returns(<<~BODY)
+        {
+          "responseCode": 400,
+          "message": "Bad request (\/form-id-submissions) - Submissions couldn't inserted",
```
Comment (breville) [inline dashboard/test/lib/services/csta_enrollment_test.rb:100]: nit: "couldn't inserted"? :)

Comment (islemaster) [reply dashboard/test/lib/services/csta_enrollment_test.rb:100]: :man_shrugging: That's exactly what their server responded with.

image

Comment (breville) [reply dashboard/test/lib/services/csta_enrollment_test.rb:100]: Oh!

**Review by breville — APPROVED** [review]

**Pull request merged**